qdax/core/emitters/qpg_emitter.py

```diff
@@ -0,0 +1,541 @@
+            greedy_policy_params,
+            critic_params,
+            transitions,
+        )
+        (
+            policy_updates,
+            policy_optimizer_state,
+        ) = self._greedy_policy_optimizer.update(
+            policy_gradient, greedy_policy_opt_state
+        )
+        greedy_policy_params = optax.apply_updates(greedy_policy_params, policy_updates)
+        # Soft update of target greedy policy
+        target_greedy_policy_params = jax.tree_map(
+            lambda x1, x2: (1.0 - self._config.soft_tau_update) * x1
+            + self._config.soft_tau_update * x2,
+            target_greedy_policy_params,
+            greedy_policy_params,
+        )
+
+        return (
+            policy_optimizer_state,
+            greedy_policy_params,
+            target_greedy_policy_params,
+        )
+
+    @partial(jax.jit, static_argnames=("self",))
+    def _mutation_function_pg(
+        self,
+        controller_params: Genotype,
+        emitter_state: QualityPGEmitterState,
+    ) -> Genotype:
+        """Apply pg mutation to a policy via multiple steps of gradient descent.
+        First, update the rewards to be diversity rewards, then apply the gradient
+        steps.
+
+        Args:
+            controller_params: a controller, supposed to be a differentiable neural
+                network.
+            emitter_state: the current state of the emitter, containing among others,
+                the replay buffer, the critic.
+
+        Returns:
+            The updated params of the neural network.
+        """
+
+        # Define new controller optimizer state
+        controller_optimizer_state = self._controllers_optimizer.init(controller_params)
+
+        def scan_train_controller(
+            carry: Tuple[QualityPGEmitterState, Genotype, optax.OptState],
+            unused: Any,
+        ) -> Tuple[Tuple[QualityPGEmitterState, Genotype, optax.OptState], Any]:
+            emitter_state, controller_params, controller_optimizer_state = carry
+            (
+                new_emitter_state,
+                new_controller_params,
+                new_controller_optimizer_state,
+            ) = self._train_controller(
+                emitter_state,
+                controller_params,
+                controller_optimizer_state,
+            )
+            return (
+                new_emitter_state,
+                new_controller_params,
+                new_controller_optimizer_state,
+            ), ()
+
+        (
+            emitter_state,
+            controller_params,
+            controller_optimizer_state,
+        ), _ = jax.lax.scan(
+            scan_train_controller,
+            (emitter_state, controller_params, controller_optimizer_state),
+            (),
+            length=self._config.num_pg_training_steps,
+        )
+
+        return controller_params
+
+    @partial(jax.jit, static_argnames=("self",))
+    def _train_controller(
```

Comment:
Again since we are doing a refactor - I think controller is also confusing. Can we change this to "policy" and then in the docstring we can put "updates policy selected from the population/archive"

Reply:
ok! makes sense! 

Also, policy is more general that controller because controller is quite specific to robotic. So this is definitely a better wording!